ext/sqlite3/database.c

```diff
@@ -259,6 +261,57 @@ busy_handler(int argc, VALUE *argv, VALUE self)
     return self;
 }
 
+#ifdef HAVE_SQLITE3_PROGRESS_HANDLER
+static int
+rb_sqlite3_progress_handler(void *context)
+{
+  sqlite3RubyPtr ctx = (sqlite3RubyPtr)context;
+
+  VALUE handle = ctx->progress_handler;
+  VALUE result = rb_funcall(handle, rb_intern("call"), 0);
+
+  if (Qfalse == result) { return 1; }
+
+  return 0;
+}
+
+/* call-seq:
+ *    progress_handler([n]) { ... }
+ *    progress_handler([n,] Class.new { def call; end }.new)
+ *
+ * Register a progress handler with this database instance.
+ * This handler will be invoked periodically during a long-running query or operation.
+ * If the handler returns +false+, the operation will be interrupted; otherwise, it continues.
+ * The parameter 'n' specifies the number of SQLite virtual machine instructions between invocations.
+ * If 'n' is not provided, the default value is 1.
+ */
+static VALUE
+progress_handler(int argc, VALUE *argv, VALUE self)
+{
+  sqlite3RubyPtr ctx;
+  VALUE block, n_value;
+
+  TypedData_Get_Struct(self, sqlite3Ruby, &database_type, ctx);
+  REQUIRE_OPEN_DB(ctx);
+
+  rb_scan_args(argc, argv, "02", &n_value, &block);
+
+  int n = NIL_P(n_value) ? 1 : NUM2INT(n_value);
```

Comment:
So if I read the documentation correctly, this define the interval of SQLite VM instructions after which the callback is called. 

I'm not sure how many instructions are executed for each query, but I suspect even the most basic query at least invoke 1 instruction per row it scans, as such `1` as a default seem extremely aggressive.

Reply:
I had it set to 1000 originally, but then I thought maybe I should default to fine-grained progress reports. But I'm happy to change it back to 1000. The [Datasette](https://datasette.io/) project defaults to 1000 for its statement timeout implementation: https://github.com/simonw/datasette/blob/00e59ec461dc0150772b999c7cc15fcb9b507d58/datasette/utils/__init__.py#L182